common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java

```diff
@@ -201,16 +201,25 @@ public int size() {
         int modifier = ascending ? 1 : -1;
 
         final List<T> sorted = copyElements();
-        LOG.info("Sorting InMemoryView of item count: " + sorted.size());
-        Collections.sort(sorted, (e1, e2) -> modifier * compare(e1, e2, getter));
-        Stream<T> stream = sorted.stream();
 
-        if (first != null) {
-          stream = stream.filter(e -> modifier * compare(e, getter, first) >= 0);
-        }
+        Stream<T> stream = null;
+        if (first != null && last != null && asKey(first).compareTo(asKey(last)) == 0) {
+          // first is last, so we just want to filter the key, and don't care about sorting!
+          // This is the normal case for stage deletion -- there's no reason to sort a million
+          // summaries to pull one stage's keys on completion of a stage.
+          stream = sorted.stream().filter(e -> modifier * compare(e, getter, first) == 0);
+        } else {
+          LOG.info("Sorting InMemoryView of item count: " + sorted.size());
+          Collections.sort(sorted, (e1, e2) -> modifier * compare(e1, e2, getter));
+          stream = sorted.stream();
```

Comment:
do we need to sort at all, even first or last is no null?

Reply:
for better or worse, that appears to be the contract.  If you can show that's excessive....

Reply:
fwiw: I suspect that the ordering is required for display purposes in the job page